actor/builtin/miner/miner_test.go

```diff
@@ -490,72 +493,195 @@ func TestMinerSubmitPoStSectorSetUpdates(t *testing.T) {
+	assert.Equal(t, uint8(0), res.Receipt.ExitCode)
+	t.Run("power is 3 sectors after third submitPoSt", func(t *testing.T) {
+		expectedPower := types.NewBytesAmount(3).Mul(types.OneKiBSectorSize)
+		actualPower := requireMinerPowerFromChain(t, minerAddr, st, vms, thirdProvingPeriodStart+3, ancestors)
+		assert.Equal(t, expectedPower, actualPower)
+
+		totalStorage := requireTotalStorageFromChain(t, st, vms, thirdProvingPeriodStart+3, ancestors)
+		assert.Equal(t, totalStorageStart.Add(expectedPower), totalStorage)
 
-		// Check that state.ProvingSet is now [2] and NextDone set is []
-		mSt := requireReadMinerActorState(t, vms, st, minerAddr)
-		assert.Equal(t, types.NewIntSet(2).Values(), mSt.ProvingSet.Values())
-		assert.Equal(t, types.EmptyIntSet().Values(), mSt.NextDoneSet.Values())				
 	})
+
+	// submitPoSt over 1 sector proving set
+	res, err = th.CreateAndApplyTestMessage(t, st, vms, minerAddr, 0, fourthProvingPeriodStart+2, "submitPoSt", ancestors, []types.PoStProof{proof}, types.EmptyIntSet())
+	assert.NoError(t, err)
+	assert.NoError(t, res.ExecutionError)
+	assert.Equal(t, uint8(0), res.Receipt.ExitCode)
+
+	t.Run("power is reduced after sectors removed from proving-set", func(t *testing.T) {
+		expectedPower := types.OneKiBSectorSize
+		actualPower := requireMinerPowerFromChain(t, minerAddr, st, vms, fourthProvingPeriodStart+3, ancestors)
+		assert.Equal(t, expectedPower, actualPower)
+
+		totalStorage := requireTotalStorageFromChain(t, st, vms, fourthProvingPeriodStart+3, ancestors)
+		assert.Equal(t, totalStorageStart.Add(expectedPower), totalStorage)
+
+	})
+}
+
+type posterCommiter struct {
+	st state.Tree
+	vms vm.StorageMap
+	ancestors []types.TipSet
+	minerAddr address.Address
+	t *testing.T
+}
+
+func (pc *posterCommiter) requireCommit(blockHeight, sectorID uint64) {
+	res, err := th.CreateAndApplyTestMessage(pc.t, pc.st, pc.vms, pc.minerAddr, 0, blockHeight, "commitSector", pc.ancestors, sectorID, th.MakeCommitment(), th.MakeCommitment(), th.MakeCommitment(), th.MakeRandomBytes(types.TwoPoRepProofPartitions.ProofLen()))
+	require.NoError(pc.t, err)
+	require.NoError(pc.t, res.ExecutionError)
+	require.Equal(pc.t, uint8(0), res.Receipt.ExitCode)
+}
+
+func (pc *posterCommiter) requirePoSt(blockHeight uint64, done types.IntSet) {
+	res, err := th.CreateAndApplyTestMessage(pc.t, pc.st, pc.vms, pc.minerAddr, 0, blockHeight, "submitPoSt", pc.ancestors, []types.PoStProof{th.MakeRandomPoStProofForTest()}, done)
+	assert.NoError(pc.t, err)
+	assert.NoError(pc.t, res.ExecutionError)
+	assert.Equal(pc.t, uint8(0), res.Receipt.ExitCode)
+}
+
+func newPosterCommitter(t *testing.T, st state.Tree, vms vm.StorageMap, ancestors []types.TipSet, minerAddr address.Address ) *posterCommiter {
+	return &posterCommiter{
+		t:         t,
+		st:        st,
+		vms:       vms,
+		ancestors: ancestors,
+		minerAddr: minerAddr,
+	}
+}
+
+func TestMinerSubmitPoStSectorDoneSetUpdates(t *testing.T) {
+	tf.UnitTest(t)
```

Comment:
@acruikshank What do you think of this refactoring?  I separated out submitting and committing from reading the chain state (here I had already made a separate helper, and I'd do something similar for power).

It's almost more confusing to read without the subtests.  It feels like this test is doing too much, but at the same time we need to test longer flows like this.  Any further suggestions?

Reply:
Feel free to advise when you get here @anorht

Reply:
The general problem here is that we're attempting to test a range of different conditions by establishing them in sequence and checking those states. This is a brittle pattern and leads to long, hard-to-understand test methods. Better would be to break this into multiple tests (or subtests) that do not share state. We sometimes end up in this place when the state is tedious to establish, and I'm ok with a pragmatic approach there sometimes so long as it doesn't get out of hand. 

This is pretty close to out of hand. I suggest breaking it up.

> we need to test longer flows like this

I don't think that's a goal. We need to test lots of state transitions, but not chaining them together.

Reply:
Yeah. I agree with @anorth. This is the pattern that works for me:

1. Extract setup to a multi-return function. This can actually be defined at the start of the test.
2. Create a sub-test for each section of this test that contains an assertion (i.e. every comment in here starting with "Check").
3. Call setup in each sub-test (so the sub-tests are independent)
4. Run the minimum number of steps in each sub-test to get to the point where you can test the thing you're trying to test. Ensure that each of these steps is tested elsewhere and don't make any assertions on the intermediate steps.
5. Extract some setup steps into functions if they are repeated enough times (I usually want the total number of lines to half before i decide this is worth it).

The `posterCommitter` is still a nice refactor though.